### PR TITLE
Authorize connections within the connection manager

### PR DESCRIPTION
**Pull request by peterschwarz**
This PR introduces authorization at the connection level, instead of at the peer level.  This ensures that any connection requested or added to the connection manager is authorized before the connection is usable by the layers above (like the peer manager or the service connection manager).

It copies (for the time being) the existing authorization code from network and reworks the the authorization manager into pool that owns the connections until they have finished the authorization process. Once done, it returns the connection to the connection manager, so it can proceed as usual.

### CODE REVIEW

File: libsplinter/src/network/connection_manager/auth/handlers.rs [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+    use std::collections::VecDeque;
+    use std::sync::{Arc, Mutex};
+
+    use protobuf::Message;
+
+    use crate::protos::authorization;
+    use crate::protos::network::{NetworkMessage, NetworkMessageType};
+
+    #[test]
+    fn connect_request_dispatch() {
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/auth/handlers.rs:420]: Tests should have a) a short description; b) a description of the purpose of the test; c) the procedure to achieve (b).

File: libsplinter/src/network/connection_manager/mod.rs [hunk 2]
```diff
@@ -529,57 +569,217 @@ where
...
+        reply_sender: Sender<Result<(), ConnectionManagerError>>,
+        internal_sender: Sender<CmMessage>,
+        auth_pool: &AuthorizationPool,
+        subscribers: &mut SubscriberMap,
+    ) {
         let endpoint = connection.remote_endpoint();
-
         let id = Uuid::new_v4().to_string();
 
+        if endpoint.contains("inproc") {
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/mod.rs:580]: Why is inproc special?

File: libsplinter/src/network/connection_manager/auth/mod.rs [hunk 3]
```diff
@@ -11,3 +11,5 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+
+pub mod protocol;
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/auth/mod.rs:17]: This should probably go in crate::protocol and follow Transact's lead.

File: libsplinter/src/network/connection_manager/mod.rs (hunk 2 above)
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/mod.rs:580]: Inproc was previously special, in that it didn't require auth (it's compiled in, after all).  This has just shifted to let the connection manager decide, versus the auth layer, which should have to be concerned about what type of connection it is.

File: libsplinter/src/network/connection_manager/auth/handlers.rs (hunk 1 above)
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/auth/handlers.rs:420]: As this was taken from the previous code, it did not originally have a comment, but will add one

Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/auth/handlers.rs:427]: Fixed

Comment (vaporos) [reply libsplinter/src/network/connection_manager/auth/handlers.rs:427]: This comment was intended to apply to all the tests in the PR.

File: libsplinter/src/network/connection_manager/auth/pool.rs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    //     self
+    // }
+
+    pub fn with_prefix(mut self, prefix: String) -> Self {
+        self.prefix = Some(prefix);
+        self
+    }
+
+    pub fn build(self) -> Result<ThreadPool, ThreadPoolBuildError> {
+        let size = self.size.unwrap_or_else(|| num_cpus::get());
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/auth/pool.rs]: I've added back the below commented code, and utilize this aspect of the builder at the callsite.

File: libsplinter/src/network/connection_manager/auth/pool.rs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    size: Option<usize>,
+    prefix: Option<String>,
+}
+
+impl ThreadPoolBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    // pub fn with_size(mut self, size: usize) -> Self {
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/auth/pool.rs]: Fixed.

File: libsplinter/src/network/connection_manager/auth/mod.rs (hunk 3 above)
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/auth/mod.rs]: Moved.

File: libsplinter/src/network/connection_manager/auth/handlers.rs (hunk 1 above)
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/auth/handlers.rs:427]: These were the original comments, but I've updated them all to be more verbose. I probably wasn't clear in the commit message that this is the same code from the existing network auth code moved here.

File: libsplinter/src/network/connection_manager/auth/handlers.rs [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use protobuf::Message;
+
+use crate::network::dispatch::{
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/auth/handlers.rs:17]: I agree, but I think, in order to have a running system, we should make this change in a future PR (possible as part of, or after integrating it into splinter as a whole).

[the remark this answers is not in the document]
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/auth/handlers.rs]: Updated as part of this PR.  The connection manager takes an `Authorizer` trait

File: libsplinter/src/network/connection_manager/mod.rs (hunk 2 above)
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/mod.rs:601]: It probably would be better to add a feature where a connection could be indicate that it is authorizable, but that would require some API changes at the transport level.  Not hard - the default implementation of the function could return true.

**Pull request merged**